source/PlayerInfo.cpp

```diff
@@ -683,78 +683,45 @@ const Date &PlayerInfo::GetDate() const
 	return date;
 }
 
-
-
-// Set the date to the next day, and perform all daily actions.
-void PlayerInfo::IncrementDate()
+// Set the date, and perform all daily actions the given number of times.
+// Supports zero negative values.
+void PlayerInfo::AdvanceDate(int amount)
 {
-	++date;
-
-	// Check if any special events should happen today.
-	auto it = gameEvents.begin();
-	while(it != gameEvents.end())
+	if(!amount)
+		return;
+	else if(amount > 0)
 	{
-		if(date < it->GetDate())
-			++it;
-		else
+		while(amount--)
 		{
-			it->Apply(*this);
-			it = gameEvents.erase(it);
-		}
-	}
+			++date;
 
-	// Check if any missions have failed because of deadlines and
-	// do any daily mission actions for those that have not failed.
-	for(Mission &mission : missions)
-	{
-		if(mission.CheckDeadline(date) && mission.IsVisible())
-			Messages::Add("You failed to meet the deadline for the mission \"" + mission.Name() + "\".",
-				Messages::Importance::Highest);
-		if(!mission.IsFailed())
-			mission.Do(Mission::DAILY, *this);
-	}
+			// Check if any special events should happen today.
+			auto it = gameEvents.begin();
+			while(it != gameEvents.end() && date >= it->GetDate())
+			{
+				GameEvent event = *it;
+				event.Apply(*this);
+				it = gameEvents.erase(it);
+			}
 
-	// Check what salaries and tribute the player receives.
-	int64_t salariesIncome = accounts.SalariesIncomeTotal();
-	int64_t tributeIncome = GetTributeTotal();
-	FleetBalance b = MaintenanceAndReturns();
-	if(salariesIncome || tributeIncome || b.assetsReturns)
-	{
-		string message = "You receive ";
-		if(salariesIncome)
-			message += Format::CreditString(salariesIncome) + " salary";
-		if(salariesIncome && tributeIncome)
-		{
-			if(b.assetsReturns)
-				message += ", ";
-			else
-				message += " and ";
+			// Check if any missions have failed because of deadlines and
+			// do any daily mission actions for those that have not failed.
+			for(Mission &mission : missions)
+			{
+				if(mission.CheckDeadline(date) && mission.IsVisible())
+					Messages::Add("You failed to meet the deadline for the mission \"" + mission.Name() + "\".",
+						Messages::Importance::Highest);
+				if(!mission.IsFailed())
+					mission.Do(Mission::DAILY, *this);
+			}
+
+			DoAccounting();
```

Comment:
same thing as before but here we should definitely do it only once, like doAccounting(days) or smth because nobody will read 1k messages when going forward in time ^^'

Reply:
Nah, I can't do it all at once. Missions and stuff triggering in the middle of the thing will need to have up-to-date info on how much money you have.